zebra-chain/src/keys/sapling.rs

```diff
@@ -7,17 +7,188 @@
 //!
 //! [ps]: https://zips.z.cash/protocol/protocol.pdf#saplingkeycomponents
 
-use std::{
-    fmt,
-    io::{self},
-};
+use std::fmt;
+
+use blake2::{Blake2b, Digest};
+use rand_core::{CryptoRng, RngCore};
 
 #[cfg(test)]
-use proptest::{array, collection::vec, prelude::*};
+use proptest::prelude::*;
 #[cfg(test)]
 use proptest_derive::Arbitrary;
 
-use crate::serialization::{SerializationError, ZcashDeserialize, ZcashSerialize};
+/// Our root secret key of the Sprout key derivation tree.
+///
+/// All other Sprout key types derive from the SpendingKey value.
+///
+#[derive(Copy, Clone, Debug, Eq, PartialEq)]
+pub struct SpendingKey(pub [u8; 32]);
+
+impl SpendingKey {
+    /// Generate a new _SpendingKey_.
+    pub fn new<T>(csprng: &mut T) -> Self
+    where
+        T: RngCore + CryptoRng,
+    {
+        let mut bytes = [0u8; 32];
+        csprng.fill_bytes(&mut bytes);
+
+        Self::from(bytes)
+    }
+}
+
+impl From<[u8; 32]> for SpendingKey {
+    /// Generate a _SpendingKey_ from existing bytes.
+    fn from(mut bytes: [u8; 32]) -> SpendingKey {
+        SpendingKey(bytes)
+    }
+}
+
+/// Derived from a _SpendingKey_.
+pub type SpendAuthorizationKey = redjubjub::Scalar;
+
+impl From<SpendingKey> for SpendAuthorizationKey {
+    /// Invokes Blake2b-512 as PRF^expand to derive a
+    /// SpendAuthorizationKey from a SpendingKey.
+    ///
+    /// https://zips.z.cash/protocol/protocol.pdf#saplingkeycomponents
+    /// https://zips.z.cash/protocol/protocol.pdf#concreteprfs
+    fn from(spending_key: SpendingKey) -> SpendAuthorizationKey {
+        let mut block = [0u8; 33]; // Last byte is t=0;
+
+        block[0..32].copy_from_slice(&spending_key.0[..]);
+
+        let mut hasher = Blake2b::new();
+        // TODO: check that this counts as personalization.
```

Comment:
Nope, personalization is a separate input.

Reply:
It is in the PRF definition, in this function there is no param for 'personalization string' so I have to figure out if concatenating it at the beginning of the preimage input is fine or if I have to find another implementation.

Reply:
It's not equivalent to prepending anything, it affects the parameter block. See section 2.8 of http://blake2.net/blake2.pdf .

Reply:
zcashd uses blake2b_simd and blake2s_simd, btw.